authority/provisioner/k8sSA.go

```diff
@@ -0,0 +1,245 @@
+			}
+			if rvw.Status.Error != "" {
+				return nil, errors.Errorf("error from kubernetes TokenReviewAPI: %s", rvw.Status.Error)
+			}
+			if !rvw.Status.Authenticated {
+				return nil, errors.New("error from kubernetes TokenReviewAPI: token could not be authenticated")
+			}
+			if err = jwt.UnsafeClaimsWithoutVerification(&claims); err != nil {
+				return nil, errors.Wrap(err, "error parsing claims")
+			}
+		*/
+	}
+	for _, pk := range p.pubKeys {
+		if err = jwt.Claims(pk, &claims); err == nil {
+			valid = true
+			break
+		}
+	}
+	if !valid {
+		return nil, errors.New("error validating token and extracting claims")
+	}
+
+	// According to "rfc7519 JSON Web Token" acceptable skew should be no
+	// more than a few minutes.
+	if err = claims.Validate(jose.Expected{
+		Issuer: k8sSAIssuer,
+	}); err != nil {
+		return nil, errors.Wrapf(err, "invalid token claims")
+	}
+
+	if claims.Subject == "" {
+		return nil, errors.New("token subject cannot be empty")
+	}
+
+	return &claims, nil
+}
+
+// AuthorizeRevoke returns an error if the provisioner does not have rights to
+// revoke the certificate with serial number in the `sub` property.
+func (p *K8sSA) AuthorizeRevoke(token string) error {
+	_, err := p.authorizeToken(token, p.audiences.Revoke)
+	return err
+}
+
+// AuthorizeSign validates the given token.
+func (p *K8sSA) AuthorizeSign(ctx context.Context, token string) ([]SignOption, error) {
+	_, err := p.authorizeToken(token, p.audiences.Sign)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check for SSH sign-ing request.
+	if MethodFromContext(ctx) == SignSSHMethod {
+		return nil, errors.New("ssh certificates not enabled for k8s ServiceAccount provisioners")
+	}
+
+	//dnsNames, ips, emails := x509util.SplitSANs(claims.SANs)
+	return []SignOption{
+		// modifiers / withOptions
+		newProvisionerExtensionOption(TypeK8sSA, p.Name, ""),
+		profileDefaultDuration(p.claimer.DefaultTLSCertDuration()),
+		// validators
+		defaultPublicKeyValidator{},
+		newValidityValidator(p.claimer.MinTLSCertDuration(), p.claimer.MaxTLSCertDuration()),
+	}, nil
+}
+
+// AuthorizeRenewal returns an error if the renewal is disabled.
+func (p *K8sSA) AuthorizeRenewal(cert *x509.Certificate) error {
+	if p.claimer.IsDisableRenewal() {
+		return errors.Errorf("renew is disabled for provisioner %s", p.GetID())
+	}
+	return nil
+}
+
+/*
+func checkAccess(authz kauthz.AuthorizationV1Interface) error {
+	r := &kauthzApi.SelfSubjectAccessReview{
+		Spec: kauthzApi.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &kauthzApi.ResourceAttributes{
+				Group:    "authentication.k8s.io",
+				Version:  "v1",
+				Resource: "tokenreviews",
+				Verb:     "create",
+			},
+		},
+	}
+	rvw, err := authz.SelfSubjectAccessReviews().Create(r)
+	if err != nil {
+		return err
+	}
+	if !rvw.Status.Allowed {
+		return fmt.Errorf("Unable to create kubernetes token reviews: %s", rvw.Status.Reason)
+	}
+
+	return nil
+}
+*/
```

Comment:
yea, so this code is sort of a guide for how we would integrate against the kubernetes API server for V2. Instead of asking users to supply the pubKeys when the provisioner is added, you would have configuration details for connection to the kubernetes API server. Then you would submit the token token to the API and it would return valid/invalid. But for now we're doing something more basic. I can leave a comment to this effect in the code.